cpu/stm32_common/periph/usbdev.c

```diff
@@ -0,0 +1,1151 @@
+/*
+ * Copyright (C) 2019 Koen Zandberg
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup cpu_stm32_common_usbdev
+ * @{
+ * @file
+ * @brief   Low level USB interface functions for the stm32 FS/HS devices
+ *
+ * @author  Koen Zandberg <koen@bergzand.net>
+ * @}
+ */
+
+#include <stdint.h>
+#include <stdlib.h>
+#include <errno.h>
+
+#include "bitarithm.h"
+#include "xtimer.h"
+#include "cpu.h"
+#include "cpu_conf.h"
+#include "periph/pm.h"
+#include "periph/gpio.h"
+#include "periph/usbdev.h"
+#include "usbdev_stm32.h"
+
+/**
+ * Be careful with enabling debug here. As with all timing critical systems it
+ * is able to interfere with USB functionality and you might see different
+ * errors than debug disabled
+ */
+#define ENABLE_DEBUG (0)
+#include "debug.h"
+
+#if defined(STM32_USB_OTG_FS_ENABLED) && defined(STM32_USB_OTG_HS_ENABLED)
+#define _TOTAL_NUM_ENDPOINTS  (STM32_USB_OTG_FS_NUM_EP + \
+                               STM32_USB_OTG_HS_NUM_EP)
+#elif defined(STM32_USB_OTG_FS_ENABLED)
+#define _TOTAL_NUM_ENDPOINTS  (STM32_USB_OTG_FS_NUM_EP)
+#elif defined(STM32_USB_OTG_HS_ENABLED)
+#define _TOTAL_NUM_ENDPOINTS  (STM32_USB_OTG_HS_NUM_EP)
+#endif
+
+/* Mask for the set of interrupts used */
+#define STM32_FSHS_USB_GINT_MASK    \
+    (USB_OTG_GINTMSK_USBSUSPM | \
+     USB_OTG_GINTMSK_WUIM     | \
+     USB_OTG_GINTMSK_ENUMDNEM | \
+     USB_OTG_GINTMSK_USBRST   | \
+     USB_OTG_GINTMSK_OTGINT   | \
+     USB_OTG_GINTMSK_IEPINT   | \
+     USB_OTG_GINTMSK_OEPINT   | \
+     USB_OTG_GINTMSK_RXFLVLM)
+
+#define STM32_PKTSTS_GONAK          0x01    /**< Rx fifo global out nak */
+#define STM32_PKTSTS_DATA_UPDT      0x02    /**< Rx fifo data update    */
+#define STM32_PKTSTS_XFER_COMP      0x03    /**< Rx fifo data complete  */
+#define STM32_PKTSTS_SETUP_COMP     0x04    /**< Rx fifo setup complete */
+#define STM32_PKTSTS_SETUP_UPDT     0x06    /**< Rx fifo setup update   */
+
+/* Some device families forgot to define the FS device FIFO size */
+#ifndef USB_OTG_FS_TOTAL_FIFO_SIZE
+#define USB_OTG_FS_TOTAL_FIFO_SIZE      (1280U)
```

Comment:
1280 ? Isn't a weird value ?

Reply:
[I don't think it is that weird](https://github.com/RIOT-OS/RIOT/blob/master/cpu/stm32f4/include/vendor/stm32f401xe.h#L8493)

Reply:
How about use the vendor define then?

Reply:
Some of the ST vendor headers (for the parts that do have this peripheral) are missing this define.

Reply:
But are we 100% sure, this value will remain the same across families ?

Reply:
Probably not, but I don't really see a different solution. I have already wrapped it in an `#ifndef` to allow the vendor headers to override the value with their own. I think that's the best we can do for now.

Reply:
The OTG fs peripheral is present, but the FIFO size define is missing in the following device vendor headers:
```
cpu/stm32f7/include/vendor/stm32f722xx.h
cpu/stm32f7/include/vendor/stm32f723xx.h
cpu/stm32f7/include/vendor/stm32f746xx.h
cpu/stm32f7/include/vendor/stm32f767xx.h
cpu/stm32f7/include/vendor/stm32f769xx.h
cpu/stm32l4/include/vendor/stm32l475xx.h
cpu/stm32l4/include/vendor/stm32l476xx.h
cpu/stm32l4/include/vendor/stm32l496xx.h
cpu/stm32l4/include/vendor/stm32l4r5xx.h
```